### ISSUE TITLE
Add ability to grant access to a registered custom protocol

### ISSUE THREAD

**Issue by RichardSteele** [MicrosoftEdge/WebView2Feedback#137]
Trying to access a registered custom protocol prompts the user for confirmation every time. It would be useful to have a way to confirm access programmatically. The PermissionRequested event might be a way, but it would need an additional member for the requested URI.

[AB#34160054](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/34160054)

**Comment by david-risney** [comment]
Thanks for the feedback. Can you elaborate on what your app does with custom protocols? Also, when you say registered custom protocol which do you mean? 

1. [Registering a Windows app to a URI scheme](https://docs.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/platform-apis/aa767914(v=vs.85)?redirectedfrom=MSDN) where you can have URIs of a registered scheme open a registered Windows app.
2. [navigator.registerProtocolHandler](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler) where you can have a website register to handle a particular URI scheme.
3. [NavigateToStreamAsync](https://docs.microsoft.com/en-us/uwp/api/windows.web.iuritostreamresolver.uritostreamasync#Windows_Web_IUriToStreamResolver_UriToStreamAsync_Windows_Foundation_Uri_) like we had in the EdgeHtml WebView1 control that would let you register your own custom code that gets to turn URIs of a set URI scheme into a Stream.

Or perhaps something else?

**Comment by RichardSteele** [comment]
Our client application integrates various third-party web applications using Microsoft's IE-based [WebBrowser control](https://docs.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/platform-apis/aa752040(v=vs.85)?redirectedfrom=MSDN). The control is customized via [internet feature control keys](https://docs.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/general-info/ee330720(v=vs.85)) to be "as modern" as possible. Unfortunately but understandably, some of our (potential) partners are unwilling to support IE11 (any longer).

There are two ways a web application can interact with our client.
The first way is to invoke a [custom URI scheme](https://docs.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/platform-apis/aa767914(v=vs.85)), triggering another application of ours, which in turn communicates with our client by Winsock. This way is limited to simple data exchange.
The second way is to use `window.external`. By implementing `IDocHostUIHandler::GetExternal`, we are able to provide a rich interface to our client.

The custom protocol predates `window.external` and isn't our preferred way. Because some web applications don't support IE11, we are forced to open them in whatever web browser a user has installed. The protocol still enables them to communicate with our client.
By changing over to WebView2, we are able to host all web applications within our client again. As long as we'll be left with protocol usage, it would be nice if a user doesn't get constantly "pestered" :).

**Comment by vmeganathan81** [comment]
@david-risney , even though you register the protocol we do get warning Popup to end users, we need to go and change the preference file of user profile that gets created for any chrome based browser in-order to suppress the pop-up. Since here we are using webview2 control, there should be option not to show the alert or to intercept any external URL schema and  re-direct  to appropriate application.

**Comment by vmeganathan81** [comment]
@RichardSteele, how did you address with webview2, is there any workable interim solution?

**Comment by vmeganathan81** [comment]
@david-risney, you have any update on this feature? Do you know if any workable solution available ?

**Comment by RichardSteele** [comment]
@vmeganathan81 I've managed with event handlers attached to NavigationStarting and FrameNavigationStarting.

**Comment by Curve** [comment]
Is there any progress on this?
Even adding the possibility to get the uri of the resource that is requested from the `ICoreWebView2WebResourceRequestedEventHandler` would be a huge step forward.

Im trying to intercept some of the resources that are requested and replace their content - this is not quite possible with the current state of things. I can successfully replace the content of **every** resource that is requested but I can't replace the content of specific resources.
In short: Would it be possible to add a function to get the uri that is requested from the `ICoreWebView2WebResourceRequestedEventArgs`?

**Comment by novac42** [comment]
This feature request has been completed and available in Canary 115.0.1831.0 and above. Thanks!
